### DISCUSSION TITLE
[Feature] Provide feature flags to enable specific plugins

### DISCUSSION THREAD

**Discussion by namit-chandwani** [aerogear/charmil#81] [Ideas]
# Description
- Feature flagging is a technique used by developers to activate or deactivate specific features in their projects.

- One of the most common use cases of feature flags is in the feature release process, with the help of which developers can offer certain users access to features before releasing them for everyone (includes early access programs, canary releases, and beta programs).

- In our case, through feature flags, we'll be allowing the SDK users to activate certain plugins in their CLI.

# Proposed Solution

- **SDK-side code:**
	```go
	var featureFlagRegistry = make(map[string][]*cobra.Command)

	type Command cobra.Command

	func (cmd *Command) CreateFeatureFlag(name, shorthand, description string) error {
		featureFlagRegistry[name] = []*cobra.Command{}

		var featFlag bool
		cmd.Flags().BoolVarP(&featFlag, name, shorthand, false, description) // Can be integrated with Viper to manage config
	}

	func AddCmdToFeatureFlag(flagName string, cmd *cobra.Command) error {
		featureFlagRegistry[flagName] = append(featureFlagRegistry[flagName], cmd)
	}
	```
	&nbsp;
	
- **HostCLI-side code:**
	```go
	func main() {
		...
		...
		...

		cmd.CreateFeatureFlag("earlyaccess")

		calculatorCmd, err := calculator.CalculatorCommand()
		if err != nil {
			fmt.Println(err)
		}

		AddCmdToFeatureFlag("earlyaccess", calculatorCmd)

		// Add code which checks the value of `earlyaccess` flag and adds commands to the root CLI based on it
	}
	```
	&nbsp;
	**PS:** This is just a minimal code example that throws some light on my approach. There might be some errors over here and many extra things yet to be added. 
	The idea is to just discuss and finalize the working of feature flags before putting them into action.
&nbsp;

# How the end result might look like

1. **Before:**
	```bash
	$ ./host -h
		
	Host CLI created using Charmil SDK

	Usage:
	  Host [command]

	Available Commands:
	  help        Help about any command

	Flags:
	  -e, --earlyaccess   Enable early access commands
	  -h, --help          help for Host
	```
2. **Enabling the early access feature flag:**
	```bash
	$ ./host --earlyaccess=true
	```

3. **After:**
	```bash
	$ ./host -h
	
	Host CLI created using Charmil SDK

	Usage:
	  Host [command]

	Available Commands:
	  calculator        Calculator CLI
	  help        		Help about any command

	Flags:
	  -e, --earlyaccess   Enable early access commands
	  -h, --help          help for Host
	  ```
  
&nbsp;
# Elements that require further research
- [ ] A system to allow only specific CLI end-users to access the feature flags (some sort of Profile-based filtering)

&nbsp;
---
CC @aerogear/charmil-core

**Comment by jackdelahunt** [comment]
I like this, being able to allow the host cli to flag a command as early access or any other version is a great feature. My only question is what about extending this to other parts of the command like the flags. Do we want to be able to mark flags as early access and if so how would we want the host cli to do that.

**Reply by wtrocki** [reply]
Cobra has way to hide command from help etc. all we need is host/global variable to define access levels and then act on them in CLI. Not
